upki-mirror/src/main.rs

```diff
@@ -1,3 +1,156 @@
-fn main() {
-    println!("Hello, world!");
+use core::time::Duration;
+use std::collections::HashMap;
+use std::fs;
+use std::time::SystemTime;
+
+use eyre::{Context, Report, bail};
+use serde::Serialize;
+use sha2::{Digest, Sha256};
+
+mod mozilla;
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() -> Result<(), Report> {
+    process(&MOZILLA_PROD).await
+}
+
+async fn process(source: &Source) -> Result<(), Report> {
+    let client = reqwest::Client::builder()
+        .use_rustls_tls()
+        .timeout(Duration::from_secs(10))
+        .user_agent(format!(
+            "{} v{} ({})",
+            env!("CARGO_PKG_NAME"),
+            env!("CARGO_PKG_VERSION"),
+            env!("CARGO_PKG_REPOSITORY")
+        ))
+        .build()
+        .wrap_err("Failed to create HTTP client")?;
+
+    let response = client
+        .get(source.records_url)
+        .send()
+        .await
+        .wrap_err("Records request failed")?;
+
+    if !response.status().is_success() {
+        bail!("HTTP error for records request: {}", response.status());
+    }
+
+    let moz_manifest = response
+        .json::<mozilla::Manifest>()
+        .await
+        .wrap_err("Failed to parse records JSON")?;
+
+    let by_parent: HashMap<String, mozilla::Item> = HashMap::from_iter(
+        moz_manifest
+            .data
+            .iter()
+            .filter_map(|it| Some((it.parent.as_ref()?.clone(), it.clone()))),
+    );
+
+    // Walk the DAG of filters, starting from the root full filter.
+    let mut next = moz_manifest.data.iter().find(|it| {
+        !it.incremental && it.parent.is_none() && it.channel == mozilla::Channel::Default
+    });
+
+    let mut download_plan = Vec::new();
+
+    while next.is_some() {
+        let item = next.unwrap();
+        next = by_parent.get(&item.id);
+        download_plan.push(item);
+    }
+
+    let mut manifest = Manifest {
+        generated_at: SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap()
+            .as_secs(),
+        comment: "hello world".into(),
+        filters: vec![],
+    };
```

Comment:
Nit: suggest moving this down below the loop, and only having `let filters = Vec::new();` here. Maybe even `with_capacity()` with some reasonable minimum size to skip a few resizes?